doc/user/content/sql/subscribe.md

```diff
@@ -288,6 +296,71 @@ If your row has a unique column key, it is possible to map the update to its cor
 
 In the example above, `Column 1` acts as the column key that uniquely identifies the origin row the update refers to; in case this was unknown, hashing the values from `Column 1` to `Column N` would identify the origin row.
 
+### Alternate output formats
+
+{{< alpha />}}
+
+#### `ENVELOPE UPSERT`
+
+Takes a list of columns to interpret as a key (given by the `KEY` option) and within each distinct timestamp interprets the updates as a series of upserts. The output columns are reordered as `(mz_timestamp, mz_state, k1, k2, .., v1, v2, ..)`. There is no `mz_diff` column. The values `v1, v2, ..` are set the last value if there's an update or an insert and are all set to `NULL` if the only updates to the key in that timestamp were deletions. `mz_state` is either "delete" or "upsert" to distinguish the two cases; when there are no values or the values can all be NULL this column is the only way to distinguish a deletion from an upsert. If `PROGRESS` is set we also return the `mz_progressed` column as usual before `mz_state` and each progress message nulls out all the key and value columns.
+
+Using `ENVELOPE UPSERT` when there is more than one live value per key can be confusing. The upserts generated have no way to express which value was deleted for a given key.
```

Comment:
I think we should encourage very strongly users to specify a key that is unique, such that this won't happen, and also how to detect if it does so they can fix their key spec. I'm less interested in seeing an example of what happens in this case, and would prefer a few sentences describing exactly the properties of the output.

Reply:
yeah, added the key violation idea to the docs + code